### DISCUSSION TITLE
MIDI Virtual Sources and Destinations

### DISCUSSION THREAD

**Discussion by Badsheep** [Hammerspoon/hammerspoon#3379] [Q&A]
Hello,

I'm currently trying to write a script to send MIDI commands (control change) to an external application.
I'm a beginner and I'm a bit lost.

The external application (Max, from Cycling 74) creates automatically 2 virtuals MIDI outputs ("From Max 1" and "From Max 2") and 2 virtual MIDI inputs ("To Max 1" and "To Max 2").

I know I need to send midi command to "To Max 1".
But I don't know how to access to that virtual MIDI canal in Hammerspoon...

When I type in the console hs.inspect(hs.midi.virtualSources()), I get { "from Max 1", "from Max 2" }.

But has its names says, these are Sources, no Destinations.
Can I access to MIDI destinations ?
Thanks.

**Comment by Badsheep** [comment]
I found a workaround.
I created a midi virtual canal in "gestionnaire IAC" (sorry it's the French name... you know what I mean, the natives midi utility in Mac OS). It's seen as a "device" (not a virtual source) : 

hs.inspect(hs.midi.devices())
{ "Réseau", "Gestionnaire IAC", "Bluetooth" }

I set that Midi port to the input of Max, and then I can send some commands using : 

toMax = hs.midi.new(hs.midi.devices()[2])
toMax:sendCommand("controlChange", {
  ["controllerNumber"] = 4,
  ["controllerValue"] = 11111111111111,
  ["channel"] = 0,
}) 

And it works :)

Still that would be nice to add the ability to see the Midi Virtual Destinations in the hs.midi API.
I'll do a Feature Request.
Thanks to all the contributors of Hammerspoon and all the API, it's really wonderful.

**Reply by latenitefilms** [reply]
I'll have to have a proper think about this at some point. I do remember with CommandPost, I needed to some [funky things](https://github.com/CommandPost/CommandPost-UserGuide/blob/854c3c2b3a2f871cea6a9ee9581a8fda73638295/midi/logic/README.md#iac-driver-workaround) with IAC to work around some problems too.

FWIW - `hs.midi` is basically just a wrapper for [MIKMIDI](https://github.com/mixedinkey-opensource/MIKMIDI).

**Reply by Vladistone** [reply]
> I'll have to have a proper think about this at some point. I do remember with CommandPost, I needed to some [funky things](https://github.com/CommandPost/CommandPost-UserGuide/blob/854c3c2b3a2f871cea6a9ee9581a8fda73638295/midi/logic/README.md#iac-driver-workaround) with IAC to work around some problems too.
> 
> FWIW - `hs.midi` is basically just a wrapper for [MIKMIDI](https://github.com/mixedinkey-opensource/MIKMIDI).

Hi there!
There is news looking for a solution see the Midi Virtual Destinations in the hs.midi API?

**Comment by Badsheep** [comment]
OK thanks !
